src/MSBuild/LiveLogger/LiveLogger.cs

```diff
@@ -42,302 +129,608 @@ internal class LiveLogger : ILogger
-            TerminalBuffer.FinishedProjects = this.finishedProjects;
+    /// <summary>
+    /// The <see cref="IEventSource.ProjectFinished"/> callback.
+    /// </summary>
+    private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
+    {
+        var buildEventContext = e.BuildEventContext;
+        if (buildEventContext is null)
+        {
+            return;
+        }
 
-            // First render
-            TerminalBuffer.Render();
-            int i = 0;
-            // Rerender periodically
-            while (!TerminalBuffer.IsTerminated)
+        // Mark node idle until something uses it again
+        if (_restoreContext is null)
+        {
+            lock (_lock)
             {
-                i++;
-                // Delay by 1/60 seconds
-                // Use task delay to avoid blocking the task, so that keyboard input is listened continously
-                Task.Delay((i / 60) * 1_000).ContinueWith((t) =>
+                _nodes[NodeIndexForContext(buildEventContext)] = null;
+            }
+        }
+
+        ProjectContext c = new(buildEventContext);
+
+        // First check if we're done restoring.
+        if (_restoreContext is ProjectContext restoreContext && c == restoreContext)
+        {
+            lock (_lock)
+            {
+                _restoreContext = null;
+
+                Stopwatch projectStopwatch = _projects[restoreContext].Stopwatch;
+                double duration = projectStopwatch.Elapsed.TotalSeconds;
+                projectStopwatch.Stop();
+
+                Terminal.BeginUpdate();
+                try
+                {
+                    EraseNodes();
+                    Terminal.WriteLine($"Restore complete ({duration:F1}s)");
+                    DisplayNodes();
+                }
+                finally
                 {
-                    TerminalBuffer.FinishedProjects = this.finishedProjects;
+                    Terminal.EndUpdate();
+                }
+                return;
+            }
+        }
+
+        // If this was a notable project build, we print it as completed only if it's produced an output or warnings/error.
+        if (_projects.TryGetValue(c, out Project? project) && (project.OutputPath is not null || project.BuildMessages is not null))
+        {
+            lock (_lock)
+            {
+                Terminal.BeginUpdate();
+                try
+                {
+                    EraseNodes();
+
+                    double duration = project.Stopwatch.Elapsed.TotalSeconds;
+                    ReadOnlyMemory<char>? outputPath = project.OutputPath;
 
-                    // Rerender projects only when needed
-                    foreach (var project in projects)
+                    Terminal.Write(Indentation);
+
+                    if (e.ProjectFile is not null)
+                    {
+                        string projectFile = Path.GetFileName(e.ProjectFile);
```

Comment:
The old-school message is `FileNameWithoutExtension`. I think I'd prefer to keep that.